### PR TITLE
feat(data): update code examples for optimistic UI docs - update types, update return value for mutation hooks, and add additional cache invalidation

### PR DESCRIPTION
**Pull request by david-mcafee**
#### Description of changes:
There were a few TypeScript errors in the optimistic UI docs that my editor didn't display until after a restart. This PR: 
1) updates types, adds extra type guards, and removes unnecessary types,
2) updates the return value for mutation hooks, which as a result required additional cache invalidation for the list query.

### Instructions

**If this PR should not be merged upon approval for any reason, please submit as a DRAFT**

Which product(s) are affected by this PR (if applicable)?
- [ ] amplify-cli
- [ ] amplify-ui
- [ ] amplify-studio
- [ ] amplify-hosting
- [x] amplify-libraries

Which platform(s) are affected by this PR (if applicable)?
- [x] JS
- [ ] iOS
- [ ] Android
- [ ] Flutter
- [ ] React Native

**Please add the product(s)/platform(s) affected to the PR title**

#### Checks

- [x] Does this PR conform to [the styleguide](https://github.com/aws-amplify/docs/blob/main/STYLEGUIDE.md)?

- [ ] Does this PR include filetypes other than markdown or images? Please add or update unit tests accordingly.

- [ ] Are any files being deleted with this PR? If so, have the needed redirects been created?

- [ ] Are all links in MDX files using the MDX link syntax rather than HTML link syntax? <br /> 
      _ref: MDX: `[link](https://link.com)` 
            HTML: `<a href="https://link.com">link</a>`_

### When this PR is ready to merge, please check the box below
- [x] Ready to merge

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx [hunk 1]
```diff
@@ -1461,7 +1521,8 @@ function App() {
                 "Updating Real Estate Property..."
               ) : (
                 <>
-                  {updateMutation.isError ? (
+                  {updateMutation.isError &&
+                  updateMutation.error instanceof Error ? (
```
Comment (iartemiev) [inline src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:1525]: For these and the next 2 - is it possible for `isError` to be true end `error` to be undefined and or not an instance of `Error`?

Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:1525]: This was the suggested approach from the TanStack docs for dealing with the resulting TypeScript errors: https://tanstack.com/query/latest/docs/react/typescript#typing-the-error-field.

Comment (iartemiev) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:1525]: Makes sense. Should `deleteMutation` and `createMutation` below follow the same pattern then?

Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:1525]: @iartemiev - The TS examples for `deleteMutation` and `createMutation` currently follow this pattern, as well.

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx [hunk 2]
```diff
@@ -158,19 +158,19 @@ const createMutation = useMutation({
...
     // Snapshot the previous value
-    const previousRealEstateProperties = queryClient.getQueryData<any>([
+    const previousRealEstateProperties = queryClient.getQueryData([
       "realEstateProperties",
     ]);
 
     // Optimistically update to the new value
     if (previousRealEstateProperties) {
-      queryClient.setQueryData<any>(["realEstateProperties"], (old: any) => [
+      queryClient.setQueryData(["realEstateProperties"], (old: any) => [
```
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:173]: Is there a way to type `old` in a better way than any?

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx [hunk 3]
```diff
@@ -328,44 +328,45 @@ const updateMutation = useMutation({
...
-    queryKey: ["realEstateProperties", newRealEstateProperty.id],
-  });
+  onMutate: async (newRealEstateProperty) => {
+    // Cancel any outgoing refetches
+    // (so they don't overwrite our optimistic update)
+    await queryClient.cancelQueries({
+      queryKey: ["realEstateProperties", newRealEstateProperty.id],
+    });
+
+    await queryClient.cancelQueries({
```
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:348]: is this case covered in the 344?

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx [hunk 4]
```diff
@@ -374,17 +375,22 @@ const updateMutation = useMutation({
...
   // Always refetch after error or success:
   onSettled: (newRealEstateProperty) => {
-    queryClient.invalidateQueries({
-      queryKey: ["realEstateProperties", newRealEstateProperty.id],
-    });
+    if (newRealEstateProperty) {
+      queryClient.invalidateQueries({
+        queryKey: ["realEstateProperties", newRealEstateProperty.id],
+      });
+      queryClient.invalidateQueries({
```
Comment (renebrandel) [inline src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:390]: is this case covered in 387?

**Review by iartemiev — APPROVED** [review]

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx (hunk 2 above)
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:173]: I tried to provide a better type here, but had some difficulty and decided to timebox the effort. If you like, I can add something to the backlog to revisit if there's something better we can do here.

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx (hunk 3 above)
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:348]: Sorry, can you clarify your question?

Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:348]: Both cancellations here are required - one invalidates the cache for the list view, the other invalidates the cache for the detail view.

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx (hunk 4 above)
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:390]: Both invalidations here are required - one invalidates the cache for the list view, the other invalidates the cache for the detail view.

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx [hunk 5]
```diff
@@ -410,6 +416,10 @@ const updateMutation = useMutation({
       queryKey: ["realEstateProperties", newRealEstateProperty.id],
     });
 
+    await queryClient.cancelQueries({
```
[the remark this answers is not in the document]
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:419]: Answers provided above. TLDR: we need to invalidate the cache for both list and detail view.

**Review by renebrandel — APPROVED** [review]

**Review by timngyn — APPROVED** [review]

File: src/fragments/lib/graphqlapi/js/optimistic-ui.mdx (hunk 2 above)
Comment (david-mcafee) [reply src/fragments/lib/graphqlapi/js/optimistic-ui.mdx:173]: Discussed offline

**Pull request merged**